docs/source/notes/distributed_autograd.rst

```diff
@@ -0,0 +1,277 @@
+Distributed Autograd Design
+===========================
+
+This note will present the detailed design for distributed autograd and walk 
+through the internals of the same. Make sure you're familiar with 
+`Autograd Mechanics`_, 
+``torch.distributed.rpc`` and ``RRef`` before proceeding 
+(RFC: https://github.com/pytorch/pytorch/issues/23110).
+
+.. 
+  TODO: Update the link above to point to actual RPC and RRef docs.
+
+.. background: 
+
+Background
+^^^^^^^^^^
+
+Let's say you have two nodes and a very simple model partitioned across two 
+nodes. This can be implemented using ``torch.distributed.rpc`` as follows:
+
+.. code::
+
+  import torch
+  import torch.distributed.rpc as rpc
+
+  def my_add(t1, t2):
+    return torch.add(t1, t2)
+
+  # On worker 0:
+  t1 = torch.rand((3, 3), requires_grad=True)
+  t2 = torch.rand((3, 3), requires_grad=True)
+
+  # Perform some computation remotely.
+  t3 = rpc.rpc_sync("worker1", my_add, args=(t1, t2))
+
+  # Perform some computation locally based on remote result.
+  t4 = torch.rand((3, 3), requires_grad=True)
+  t5 = torch.mul(t3, t4)
+
+  # Compute some loss.
+  loss = t5.sum()
+
+The main motivation behind distributed autograd is to enable running a backward 
+pass on such distributed models with the ``loss`` that we've computed and 
+record appropriate gradients for all tensors that require gradients.
+
+.. attaching_send_recv_functions:
+
+Autograd recording during the forward pass
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+PyTorch builds the autograd graph during the forward pass and this graph is 
+used to execute the backward pass. For more details see 
+`How autograd encodes history`_. 
+
+For distributed autograd, we need to keep track of all RPCs during the forward 
+pass to ensure the backward pass is executed appropriately. For this purpose, 
+we attach ``send`` and ``recv`` functions to the autograd graph when we perform 
+an RPC.
+
+- The ``send`` function is attached to the source of the RPC and it's output 
+  edges point to the input tensors of the RPC. The input for this function is 
+  received over the wire from the appropriate ``recv`` function.
+- The ``recv`` function is attached to the destination of the RPC and its 
+  inputs are retrieved from operators executed on the destination using the 
+  input tensors. The output gradients of this function are passed over the wire 
+  to the appropriate ``send`` function during the backward pass.
+- Each ``send-recv`` pair is assigned a globally unique ``autograd_message_id`` 
+  to uniquely identify the pair. This is useful to lookup the corresponding 
+  function on a remote node during the backward pass.
+- For ``RRef``, whenever we call ``RRef.to_here()`` we attach an appropriate 
+  ``send-recv`` pair for the tensors involved.
+
+As an example, this is what the autograd graph for our example above would look 
+like (t5.sum() excluded for simplicity):
+
+.. image:: images/send_recv_functions.png
```

Comment:
This might be an issue with ghexport, if you look at the phabricator diff the images are present.

Reply:
Will this be the final folder for the images or will they be moved to https://github.com/pytorch/pytorch/tree/master/docs/source/_static/img

Reply:
@jlin27 I'm not sure what the convention here is? That folder does seem to have the  logos etc. Do we intend to have design doc specific images there? I feel you feel that folder is a better place, I'll move the images there.

Reply:
We can add a folder under .../_static/img/autograd similar to how tensorboard (.../_static/imag/tensorboard) did and as long as we reference the direct path, it will get built out in the html.

Reference image from rst in tensorboard:
https://github.com/pytorch/pytorch/blob/master/torch/utils/tensorboard/writer.py#L293